src/qibo/_openqasm.py

```diff
@@ -0,0 +1,187 @@
+import numpy as np
+import openqasm3
+
+import qibo
+from qibo.config import raise_error
+
+
+class DefinedGate:
+    """Object that handles the definition of custom gates in QASM via the `gate` command.
+
+    Args:
+        gates (list): List of gates composing the defined gate.
+        qubits (list, tuple): Qubits identifiers (e.g. (q0, q1, q2, ...)).
+        args (list, tuple): Arguments identifiers (e.g. (theta, alpha, gamma, ...)).
+    """
+
+    def __init__(self, gates, qubits, args):
+        self.gates = gates
+        self.qubits = qubits
+        self.args = args
+
+    def get_gates(self, qubits, args):
+        """Return the list of gates composing the defined gate applied on the specified qubits with the specified args.
+
+        Args:
+            qubits (list, tuple): Qubits where to apply the gates.
+            args (list, tuple): Arguments to evaluate the gates on.
+
+        Returns:
+            (list) The gates of the composed gate evaluated on the input qubits with the input arguments.
+        """
+        qubit_map = dict(zip(self.qubits, qubits))
+        args_map = dict(zip(self.args, args))
+        gates = []
+        for gate in self.gates:
+            new_qubits = [qubit_map[q] for q in gate.qubits]
+            kwargs = gate.init_kwargs
+            new_args = [args_map.get(arg, arg) for arg in kwargs.values()]
+            gates.append(gate.__class__(*new_qubits, *new_args))
+        return gates
+
+
+def _qibo_gate_name(gate):
+    if gate == "cx":
+        return "CNOT"
+    elif gate == "id":
+        return "I"
+    elif gate == "ccx":
+        return "TOFFOLI"
+    else:
+        return gate.upper()
+
+
+class QASMParser:
+    """Wrapper around the :class:`openqasm3.parser` parser for QASM 3.0."""
+
+    def __init__(
+        self,
+    ):
+        self.parser = openqasm3.parser
+        self.defined_gates = {}
+        self.q_registers, self.c_registers = {}, set()
+
+    def to_circuit(self, qasm_string: str, accelerators=None, density_matrix=False):
```

Comment:
```suggestion
    def to_circuit(self, qasm_string: str, accelerators: dict = None, density_matrix: bool = False):
```

Reply:
Also, it's missing the new `wite_names` argument from `qibo.Circuit`

Reply:
but does QASM support wire names?

Reply:
[Yes](https://github.com/Qiskit/qiskit/pull/9100)

Reply:
Ok I implemented the setting of the `wire_names` based on the names of the defined quantum registers in the qasm string, but still `wire_names` should not be passed to `to_circuit` here.